cpp/examples/protobuf/dynamic_reader.cpp

```diff
@@ -0,0 +1,90 @@
+// Example code for reading Protobuf messages from an MCAP file. Uses the proto definition
+// from within the MCAP, with no dependency on generated headers.
+// Try it out by generating some PointCloud messages with the protobuf writer example,
+// and running this executable with the resulting MCAP file.
+#define MCAP_IMPLEMENTATION
+#include <google/protobuf/descriptor.pb.h>
+#include <google/protobuf/descriptor_database.h>
+#include <google/protobuf/dynamic_message.h>
+
+#include "mcap/reader.hpp"
+#include <vector>
+
+namespace gp = google::protobuf;
+
+int main(int argc, char** argv) {
+  if (argc != 2) {
+    std::cerr << "Usage: " << argv[0] << " <input.mcap>" << std::endl;
+    return 1;
+  }
+  const char* inputFilename = argv[1];
+
+  mcap::McapReader reader;
+  {
+    const auto res = reader.open(inputFilename);
+    if (!res.ok()) {
+      std::cerr << "Failed to open " << inputFilename << " for reading: " << res.message
+                << std::endl;
+      return 1;
+    }
+  }
+
+  auto messageView = reader.readMessages();
+  gp::SimpleDescriptorDatabase protoDb;
+  gp::DescriptorPool protoPool(&protoDb);
+  gp::DynamicMessageFactory protoFactory(&protoPool);
+
+  std::cout << "topic\t\ttype\t\t\ttimestamp\t\tfields" << std::endl;
+
+  for (auto it = messageView.begin(); it != messageView.end(); it++) {
+    // skip any non-protobuf-encoded messages.
+    if (it->schema->encoding != "protobuf") {
+      continue;
+    }
+    // If the proto descriptor is not yet loaded, load it.
+    if (protoPool.FindMessageTypeByName(it->schema->name) == nullptr) {
```

Comment:
I think we might want to recommend caching the descriptors by schema id (or channel id?) rather than having to do a find by name for every single message in the file (which I am guessing would be somewhat slower). It would also help readability to break out this FileDescriptorSet parsing into a separate function.

Reply:
I'm not exactly sure how safe that is - the [DescriptorPool docs](https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.descriptor#DescriptorPool) describe each descriptor as a "whole lot of cross-linked bits of data", which makes me unsure if the pointer returned from `FindMessageTypeByName` is safe to retain between `BuildFile` calls.

Reply:
I would hope the docs would explain the usage rules if it is not safe to do... but I also don't want to read all the code to figure it out.